### PR TITLE
Improve vfsgen to not unzip bindata files but send to browser directly

### LINKED ISSUE #7107: Static embed files should not be decompress and compress
As I said on https://github.com/go-gitea/gitea/pull/7080#issuecomment-498038665 , currently we use `http.FileSystem` to open a embed file on memory, that will decompress from a gzip compressed memory file and then we serve it as a http request and compress it(maybe). So I think that we can provide the gzip compressed file directly.

From vfsgen, we could detect if the `File` is `httpgzip.NotWorthGzipCompressing` or `httpgzip.GzipByter` to do that.

### PR DESCRIPTION
**Pull request by lunny**
Improve the embedded vfsgen, should fix #7107

**UPDATE:  2020.12.23**

Currently, all bindata files stored on binary with compressed bytes, when serve for http request, Gitea creates a zipReader and unzip them for the web browser. If we enabled gzip, then the file will be compressed again and sent to the browser.

But in fact, it's unnecessary. We can just send the compressed files to browser directly. It will spend less cpu and memory than before.

### CODE REVIEW

**Review by zeripath — COMMENTED** [review]
This looks like replication of modules/gzip. How come we can't use that?

Why are we adding another external dependency?

**Comment by lunny** [comment]
> This looks like replication of modules/gzip. How come we can't use that?
> 
> Why are we adding another external dependency?

This is different. See my update on PR content.

**Review by khmarbaise — APPROVED** [review]

**Comment by silverwind** [comment]
What if the browser does not support gzip `Content-Encoding`? I think this at least needs a check for `Accept-Encoding` like I did in https://github.com/go-gitea/gitea/pull/12398/files#diff-cd0e2eab70b990fde2f1bce199926a38f33eb2cdbc8a7410adac886e213dedc7R113-R137 and if the client does not support `gzip`, send plain.

**Comment by silverwind** [comment]
Also I'd suggest dropping `shurcooL/httpgzip` because it does not support brotli and implementing content-type negotiation is like 15 lines of code like I linked. Not worth pulling a dependency in imho.

**Comment by lunny** [comment]
> Also I'd suggest dropping `shurcooL/httpgzip` because it does not support brotli and implementing `Content-Encoding` negotiation is like 15 lines of code like I linked. Not worth pulling a dependency in for.

Good idea. All done except to support `br` since vfsgen don't support that.

File: modules/public/public.go [hunk 1]
```diff
@@ -87,6 +87,16 @@ func (opts *Options) staticHandler(dir string) func(next http.Handler) http.Hand
 	}
 }
 
+// parseAcceptEncoding parse Accept-Encoding: deflate, gzip;q=1.0, *;q=0.5 as compress methods
+func parseAcceptEncoding(val string) map[string]bool {
+	parts := strings.Split(val, ";")
+	var types = make(map[string]bool)
+	for _, v := range strings.Split(parts[0], ",") {
```
Comment (silverwind) [inline modules/public/public.go:94]: What about whitespace? E.g. `Accept-Encoding: gzip, deflate, br`? A test would be nice.

**Comment by silverwind** [comment]
Maybe you can also copy the docs update regarding `ENABLE_GZIP` from https://github.com/go-gitea/gitea/pull/12398.

**Comment by lunny** [comment]
@silverwind Done.

**Comment by silverwind** [comment]
Does not seem to work for me, HTTP requests are getting stuck unanswered with a bindata build.

**Comment by lunny** [comment]
Fixed. `content.Open` will still uncompress the file. We have to get compressed file content from `*vfsgen۰CompressedFileInfo`.`GzipBytes`.

**Review by silverwind — APPROVED** [review]

**Comment by silverwind** [comment]
> content.Open will still uncompress the file

That's fine. Most clients will support gzip and will get the fast path without decompression.

**Comment by lunny** [comment]
make L-G-T-M work.

**Pull request merged**